# Load Job Data


desired columns :
title|description|minimumYearsExperience|shiftPattern|skills|otherRequirements|salary_minimum|salary_maximum|salary_type_id|salary_type|occupationId|ssocCode|ssocVersion|workingHours|numberOfVacancies|categories|employmentTypes|positionLevels|ssecEqa|ssecFos|ssicCode|ssicCode2020|

In [ ]:
from pathlib import Path
import json
import pandas as pd
import numpy as np

def extract_job_jsons_to_df(job_folder="../../data") -> pd.DataFrame:
    job_folder = Path(job_folder)
    rows = []

    files = [f for f in job_folder.rglob("*.json") if f.name.startswith("MCF-")]
    total_files = len(files)

    for i, file_path in enumerate(files, start=1):
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)

            salary = data.get("salary", {}) or {}
            salary_type = salary.get("type", {}) or {}

            row = {
                "title": data.get("title"),
                "description": data.get("description"),
                "minimumYearsExperience": data.get("minimumYearsExperience"),
                "shiftPattern": data.get("shiftPattern"),
                "skills": data.get("skills"),
                "otherRequirements": data.get("otherRequirements"),
                "salary_minimum": salary.get("minimum"),
                "salary_maximum": salary.get("maximum"),
                "salary_type_id": salary_type.get("id"),
                "salary_type": salary_type.get("salaryType"),
                "occupationId": data.get("occupationId"),
                "ssocCode": data.get("ssocCode"),
                "ssocVersion": data.get("ssocVersion"),
                # "workingHours": data.get("workingHours"), # remove, all null
                "numberOfVacancies": data.get("numberOfVacancies"),
                "categories": data.get("categories"),
                "employmentTypes": data.get("employmentTypes"),
                "positionLevels": data.get("positionLevels"),
                "ssecEqa": data.get("ssecEqa"),
                "ssecFos": data.get("ssecFos"),
                "ssicCode": data.get("ssicCode"),
                "ssicCode2020": data.get("ssicCode2020"),
            }

            rows.append(row)

        except Exception as e:
            print(f"Error reading {file_path}: {e}")

        if i % 1000 == 0:
            print(f"Processed {i}/{total_files} files ({(i/total_files)*100:.1f}%)")

    print(f"Done. Total processed: {total_files}")

    return pd.DataFrame(rows)

In [ ]:
df_jobs = extract_job_jsons_to_df("../../data")
print(f"Loaded job records: {df_jobs.shape[0]}")
print(df_jobs.head())

Processed 1000/22718 files (4.4%)
Processed 2000/22718 files (8.8%)
Processed 3000/22718 files (13.2%)
Processed 4000/22718 files (17.6%)
Processed 5000/22718 files (22.0%)
Processed 6000/22718 files (26.4%)
Processed 7000/22718 files (30.8%)
Processed 8000/22718 files (35.2%)
Processed 9000/22718 files (39.6%)
Processed 10000/22718 files (44.0%)
Processed 11000/22718 files (48.4%)
Processed 12000/22718 files (52.8%)
Processed 13000/22718 files (57.2%)
Processed 14000/22718 files (61.6%)
Processed 15000/22718 files (66.0%)
Processed 16000/22718 files (70.4%)
Processed 17000/22718 files (74.8%)
Processed 18000/22718 files (79.2%)
Processed 19000/22718 files (83.6%)
Processed 20000/22718 files (88.0%)
Processed 21000/22718 files (92.4%)
Processed 22000/22718 files (96.8%)
Done. Total processed: 22718
Loaded job records: 22718
                                               title  \
0  Outdoor Sales Engineer (Own Car) | Petrochemic...   
1  IT Administrator (Coding) @ Changi (1 Year Ren...

In [ ]:
df_jobs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22718 entries, 0 to 22717
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   title                   22718 non-null  object
 1   description             22718 non-null  object
 2   minimumYearsExperience  22718 non-null  int64 
 3   shiftPattern            0 non-null      object
 4   skills                  22718 non-null  object
 5   otherRequirements       0 non-null      object
 6   salary_minimum          22718 non-null  int64 
 7   salary_maximum          22718 non-null  int64 
 8   salary_type_id          22718 non-null  int64 
 9   salary_type             22718 non-null  object
 10  occupationId            22718 non-null  object
 11  ssocCode                22718 non-null  object
 12  ssocVersion             22718 non-null  object
 13  numberOfVacancies       22718 non-null  int64 
 14  categories              22718 non-null  object
 15  em

In [ ]:
df_jobs.head()

,title,description,minimumYearsExperience,shiftPattern,skills,otherRequirements,salary_minimum,salary_maximum,salary_type_id,salary_type,...,ssocCode,ssocVersion,numberOfVacancies,categories,employmentTypes,positionLevels,ssecEqa,ssecFos,ssicCode,ssicCode2020
0,Outdoor Sales Engineer (Own Car) | Petrochemic...,<ul><li><p><strong>Role: Outdoor Sales Enginee...,1,None,"[{'skill': 'Project Bidding', 'uuid': '029f44d...",None,3000,6000,4,Monthly,...,24331,2020v2,2,"[{'id': 11, 'category': 'Engineering'}, {'id':...","[{'id': 7, 'employmentType': 'Permanent'}, {'i...","[{'id': 9, 'position': 'Executive'}]",62,0549,None,None
1,IT Administrator (Coding) @ Changi (1 Year Ren...,<p><u><strong>Job Details:</strong></u></p><ul...,3,None,"[{'skill': 'Outlook', 'uuid': '038e648f69b23b2...",None,4500,5900,4,Monthly,...,13304,2020v2,1,"[{'id': 2, 'category': 'Admin / Secretarial'},...","[{'id': 3, 'employmentType': 'Contract'}, {'id...","[{'id': 8, 'position': 'Senior Executive'}]",51,0519,None,None
2,Outsourced Finance Executive,<p><u><strong>Job Description</strong></u></p>...,2,None,"[{'skill': 'Accounts Payable', 'uuid': '1ed711...",None,3000,4000,4,Monthly,...,43129,2020v2,1,"[{'id': 1, 'category': 'Accounting / Auditing ...","[{'id': 8, 'employmentType': 'Full Time'}]","[{'id': 9, 'position': 'Executive'}]",70,0521,None,None
3,Enterprise Architect (JD#10809),<p>Job Summary</p><p>An opportunity for an Ent...,11,None,"[{'skill': 'COBIT', 'uuid': '0afebee04b864ad4b...",None,12000,14000,4,Monthly,...,25113,2020v2,1,"[{'id': 21, 'category': 'Information Technolog...","[{'id': 7, 'employmentType': 'Permanent'}]","[{'id': 2, 'position': 'Middle Management'}]",70,0919,None,None
4,L2 Preschool Malay Language Teacher,<p><strong>Job Description:</strong></p>\n<ul>...,1,None,"[{'skill': 'Childcare', 'uuid': '08370cd014507...",None,4000,5000,4,Monthly,...,36100,2020v2,5,"[{'id': 10, 'category': 'Education and Trainin...","[{'id': 7, 'employmentType': 'Permanent'}, {'i...","[{'id': 12, 'position': 'Fresh/entry level'}]",62,0112,None,None


# Text Preprocessing  
- Cleans raw text in the description columns for both courses and jobs.
- Removes HTML tags (e.g., <p> tags from job descriptions).
- Strips extra whitespace and lowercases everything.
- Creates new columns: description_clean for both all_courses_df and df_jobs.

In [ ]:
import re
from bs4 import BeautifulSoup

def clean_text(text):
    if not isinstance(text, str):
        return ""
    text = BeautifulSoup(text, "html.parser").get_text()
    text = re.sub(r'\s+', ' ', text).strip()
    text = text.lower()
    return text

# Apply to jobs
if 'description' in df_jobs.columns:
    df_jobs['description_clean'] = df_jobs['description'].apply(clean_text)
else:
    print("Warning: 'description' column not found in df_jobs")
    df_jobs['description_clean'] = ""

print("Text preprocessing completed for jobs.")

Text preprocessing completed for courses and jobs.


# Job Data Cleaning

## Job Skills Extraction

- Focuses on the structured skills field in job data (not the text descriptions).
- Extracts unique skills from df_jobs['skills'] (a list of dicts like [{"skill": "Python"}, ...]).
- Builds a vocabulary set (skill_vocab) of all unique skills across jobs.
- Adds placeholder skills column to courses (empty lists, for later population).
- Normalizes the skills column in jobs to create skills_extracted (handles None, NaN, lists, etc.).

In [ ]:
# pip install torch keybert sentence-transformers

In [ ]:
# Simple Skill Extraction from Job Data
print("\n=== Skill Extraction ===")

# Extract skills directly from job postings
def extract_skills_from_jobs(jobs_df):
    """Extract unique skills from job postings"""
    skills_vocab = set()
    if 'skills' in jobs_df.columns:
        for skills_list in jobs_df['skills'].dropna():
            if isinstance(skills_list, list):
                for item in skills_list:
                    if isinstance(item, dict) and 'skill' in item:
                        skill = str(item['skill']).strip()
                        if skill and len(skill) > 2:
                            skills_vocab.add(skill)
    return sorted(list(skills_vocab))

skill_vocab = extract_skills_from_jobs(df_jobs)
print(f"Extracted {len(skill_vocab)} unique skills from job data")


def normalize_skills(x):
    if x is None:
        return []
    if isinstance(x, float) and pd.isna(x):
        return []
    if isinstance(x, (list, tuple, set)):
        return list(x)
    # For array-likes (e.g. pandas Series) and iterables that are not strings/dicts
    if not isinstance(x, (str, bytes, dict)):
        try:
            iterable = list(x)
            return iterable
        except TypeError:
            pass
    # If skill entries are stored as strings, preserve as single-item list
    if isinstance(x, str) and x.strip():
        return [x.strip()]
    return []

# Use normalized skills list for extracted skills
if 'skills' in df_jobs.columns:
    df_jobs['skills_extracted'] = df_jobs['skills'].apply(normalize_skills)
else:
    df_jobs['skills_extracted'] = []

print(f"Skills columns added to courses and jobs")


=== Skill Extraction ===
Extracted 9456 unique skills from job data
Skills columns added to courses and jobs


## Job Data Cleaning

We will only consider jobs that require minimumYearsExperience <= 1 (for freshgraduates)

In [ ]:
# remove data with minimumYearsExperience > 1 
df_jobs = df_jobs[df_jobs["minimumYearsExperience"] <= 1]

In [ ]:

df_jobs["minimumYearsExperience"].value_counts()

minimumYearsExperience
1    5368
0    4109
Name: count, dtype: int64

minimumYearsExperience
1    5368
0    4109
Name: count, dtype: int64
There is much less job data available, after filtering for minimumYearsExperience

Also some columns have 0 values, we will remove those columns

In [ ]:
# remove columns with 0 values
# columns_to_remove = ["shiftPattern", "otherRequirements", "ssicCode", "ssicCode2020"]
# df_jobs = df_jobs.drop(columns=columns_to_remove)

In [ ]:
# Clean positionLevels column 

df_jobs["position"] = df_jobs["positionLevels"].apply(
    lambda x: x[0].get("position") 
    if isinstance(x, list) and len(x) > 0 and isinstance(x[0], dict) 
    else None
)

In [ ]:
df_jobs.loc[:, ["positionLevels", "position"]].head()

,positionLevels,position
0,"[{'id': 9, 'position': 'Executive'}]",Executive
4,"[{'id': 12, 'position': 'Fresh/entry level'}]",Fresh/entry level
5,"[{'id': 10, 'position': 'Junior Executive'}]",Junior Executive
8,"[{'id': 12, 'position': 'Fresh/entry level'}]",Fresh/entry level
11,"[{'id': 12, 'position': 'Fresh/entry level'}]",Fresh/entry level


In [ ]:
# find all unique positions
df_jobs["position"].value_counts()


position
Fresh/entry level    3969
Non-executive        1739
Executive            1638
Junior Executive     1455
Professional          374
Senior Executive      122
Manager               117
Middle Management      37
Senior Management      26
Name: count, dtype: int64

In [ ]:
# find all data with "Senior Executive" position
# senior_executives = df_jobs[df_jobs["position"] == "Senior Executive"]
# senior_executives

there are jobs with hourly pay but this is not indicated correctly in payment type

In [ ]:
# find distribution of salary for salary_minimum 
df_jobs["salary_minimum"].describe()

count     9477.000000
mean      2813.350955
std       1175.608892
min          1.000000
25%       2200.000000
50%       2800.000000
75%       3200.000000
max      18000.000000
Name: salary_minimum, dtype: float64

In [ ]:
# find all rows of data with salary less than 5% and more than 95% of the distribution
salary_min_5 = df_jobs["salary_minimum"].quantile(0.05)
salary_min_95 = df_jobs["salary_minimum"].quantile(0.95)
df_jobs["salary_minimum"].between(salary_min_5, salary_min_95).value_counts()  

salary_minimum
True     8568
False     909
Name: count, dtype: int64


columns before cleaning :
title|description|minimumYearsExperience|shiftPattern|skills|otherRequirements|salary_minimum|salary_maximum|salary_type_id|salary_type|occupationId|ssocCode|ssocVersion|workingHours|numberOfVacancies|categories|employmentTypes|positionLevels|ssecEqa|ssecFos|ssicCode|ssicCode2020|


columns after removing the null columns :
title|description_clean|minimumYearsExperience|position|skills_extracted|salary_minimum|salary_maximum|salary_type_id|salary_type|occupationId|ssocCode|ssocVersion|numberOfVacancies|categories|employmentTypes|positionLevels|ssecEqa|ssecFos


In [ ]:
print("\n=== Job Data Cleaning ===")

# Create description_clean if it doesn't exist
if 'description_clean' not in df_jobs.columns:
    print("Creating description_clean column from description...")
    df_jobs['description_clean'] = df_jobs['description'].apply(
        lambda x: clean_text(x) if isinstance(x, str) else ""
    )

# Filter for meaningful descriptions (>=10 words)
before_clean = len(df_jobs)
valid_desc = df_jobs['description_clean'].str.split().str.len() >= 10
df_jobs = df_jobs[valid_desc]
print(f"Filtered by description length: {before_clean} -> {len(df_jobs)} records")

# Remove null titles and descriptions
before_null = len(df_jobs)
df_jobs = df_jobs.dropna(subset=['title', 'description'])
print(f"Removed null values: {before_null} -> {len(df_jobs)} records")

# Remove duplicate job postings (by title and description)
before_dup = len(df_jobs)
df_jobs = df_jobs.drop_duplicates(subset=['title', 'description'], keep='first')
print(f"Removed duplicates: {before_dup} -> {len(df_jobs)} records")

print(f"\nFinal cleaned jobs shape: {df_jobs.shape}")
print(f"Columns: {list(df_jobs.columns[:10])}...")  # Show first 10 columns

print("\n=== Job Data Cleaning Complete ===")
print("Cleaned df_jobs shape:", df_jobs.shape)

# Select desired columns for display
desired_columns = [
    "title", "description_clean", "minimumYearsExperience", "position", "skills_extracted",
    "salary_minimum", "salary_maximum", "salary_type_id", "salary_type", "occupationId",
    "ssocCode", "ssocVersion", "numberOfVacancies", "categories", "employmentTypes",
    "positionLevels", "ssecEqa", "ssecFos"
]

display(df_jobs[desired_columns].head())


=== Job Data Cleaning ===
Filtered by description length: 9477 -> 9476 records
Removed null values: 9476 -> 9476 records
Removed duplicates: 9476 -> 7674 records

Final cleaned jobs shape: (7674, 24)
Columns: ['title', 'description', 'minimumYearsExperience', 'shiftPattern', 'skills', 'otherRequirements', 'salary_minimum', 'salary_maximum', 'salary_type_id', 'salary_type']...

=== Job Data Cleaning Complete ===
Cleaned df_jobs shape: (7674, 24)


,title,description_clean,minimumYearsExperience,position,skills_extracted,salary_minimum,salary_maximum,salary_type_id,salary_type,occupationId,ssocCode,ssocVersion,numberOfVacancies,categories,employmentTypes,positionLevels,ssecEqa,ssecFos
0,Outdoor Sales Engineer (Own Car) | Petrochemic...,role: outdoor sales engineerworking hours: 8.1...,1,Executive,"[{'skill': 'Project Bidding', 'uuid': '029f44d...",3000,6000,4,Monthly,OCC003416,24331,2020v2,2,"[{'id': 11, 'category': 'Engineering'}, {'id':...","[{'id': 7, 'employmentType': 'Permanent'}, {'i...","[{'id': 9, 'position': 'Executive'}]",62,0549
4,L2 Preschool Malay Language Teacher,job description: implement age-appropriate tea...,1,Fresh/entry level,"[{'skill': 'Childcare', 'uuid': '08370cd014507...",4000,5000,4,Monthly,OCC006865,36100,2020v2,5,"[{'id': 10, 'category': 'Education and Trainin...","[{'id': 7, 'employmentType': 'Permanent'}, {'i...","[{'id': 12, 'position': 'Fresh/entry level'}]",62,0112
5,7506 - Installation Technician [ Class 3 ],working hours : 8:30am - 6pm (no ot) working d...,1,Junior Executive,"[{'skill': 'Preventive Maintenance', 'uuid': '...",2600,4500,4,Monthly,OCC008559,33499,2020v2,1,"[{'id': 41, 'category': 'Others'}]","[{'id': 7, 'employmentType': 'Permanent'}, {'i...","[{'id': 10, 'position': 'Junior Executive'}]",32,None
8,Entry Level Marketing Associate (Travel Opport...,🌟 ready to build a career you’re excited about...,0,Fresh/entry level,"[{'skill': 'Customer Relationship', 'uuid': '0...",3500,5000,4,Monthly,OCC000711,12222,2020v2,5,"[{'id': 26, 'category': 'Marketing / Public Re...","[{'id': 8, 'employmentType': 'Full Time'}]","[{'id': 12, 'position': 'Fresh/entry level'}]",32,None
11,Videographer + Social Media,seeking an independent and detail-oriented soc...,1,Fresh/entry level,"[{'skill': 'Lighting', 'uuid': '2e4b97fde8cf63...",1200,3000,4,Monthly,OCC062006,24314,2020v2,1,"[{'id': 3, 'category': 'Advertising / Media'},...","[{'id': 3, 'employmentType': 'Contract'}, {'id...","[{'id': 12, 'position': 'Fresh/entry level'}]",51,0239


In [ ]:
# clean & update 'skills_extracted', 'categories', 'employmentTypes' and 'positionLevels' columns to only contain the relevant values, remove other metadata like uuid, confidence, isKeySkill, id etc.

# Update skills_extracted to keep only 'skill' values, ignore uuid, confidence, isKeySkill
df_jobs['skills_extracted'] = df_jobs['skills_extracted'].apply(
    lambda x: ', '.join([item['skill'] for item in x if isinstance(item, dict) and 'skill' in item]) or None if isinstance(x, list) else None
)

# Update categories to only contain list of 'category', remove 'id'
df_jobs['categories'] = df_jobs['categories'].apply(
    lambda x: ', '.join([cat['category'] for cat in x if isinstance(cat, dict) and 'category' in cat]) or None if isinstance(x, list) else None
)

# Update employmentTypes to only contain list of 'employmentType', remove 'id'
df_jobs['employmentTypes'] = df_jobs['employmentTypes'].apply(
    lambda x: ', '.join([et['employmentType'] for et in x if isinstance(et, dict) and 'employmentType' in et]) or None if isinstance(x, list) else None
)

# Update positionLevels to only contain list of 'position', remove 'id'
df_jobs['positionLevels'] = df_jobs['positionLevels'].apply(
    lambda x: ', '.join([pl['position'] for pl in x if isinstance(pl, dict) and 'position' in pl]) or None if isinstance(x, list) else None
)

display(df_jobs["skills_extracted"])

0        Project Bidding, Technical Demonstrations, Shi...
4        Childcare, Written English, Classroom, Teachin...
5        Preventive Maintenance, Troubleshooting, Const...
8        Customer Relationship, Leadership, Verbal Comm...
11       Lighting, Video Editor, Property, Adobe Premie...
                               ...                        
22707    Animal Feed, Animal Care, Veterinary, Microsof...
22708    Management Skills, Producing, Leadership, Poli...
22709    Managed Care, Housekeeping, Treatment, Healthc...
22710    Auditing Standards, Risk Assessment, ACCA, Aud...
22711    Sales, Employer Branding, Microsoft Office, Ta...
Name: skills_extracted, Length: 7674, dtype: object

In [ ]:
display(df_jobs[desired_columns].head())

,title,description_clean,minimumYearsExperience,position,skills_extracted,salary_minimum,salary_maximum,salary_type_id,salary_type,occupationId,ssocCode,ssocVersion,numberOfVacancies,categories,employmentTypes,positionLevels,ssecEqa,ssecFos
0,Outdoor Sales Engineer (Own Car) | Petrochemic...,role: outdoor sales engineerworking hours: 8.1...,1,Executive,"Project Bidding, Technical Demonstrations, Shi...",3000,6000,4,Monthly,OCC003416,24331,2020v2,2,"Engineering, Manufacturing, Marketing / Public...","Permanent, Full Time",Executive,62,0549
4,L2 Preschool Malay Language Teacher,job description: implement age-appropriate tea...,1,Fresh/entry level,"Childcare, Written English, Classroom, Teachin...",4000,5000,4,Monthly,OCC006865,36100,2020v2,5,Education and Training,"Permanent, Full Time",Fresh/entry level,62,0112
5,7506 - Installation Technician [ Class 3 ],working hours : 8:30am - 6pm (no ot) working d...,1,Junior Executive,"Preventive Maintenance, Troubleshooting, Const...",2600,4500,4,Monthly,OCC008559,33499,2020v2,1,Others,"Permanent, Full Time",Junior Executive,32,None
8,Entry Level Marketing Associate (Travel Opport...,🌟 ready to build a career you’re excited about...,0,Fresh/entry level,"Customer Relationship, Leadership, Verbal Comm...",3500,5000,4,Monthly,OCC000711,12222,2020v2,5,Marketing / Public Relations,Full Time,Fresh/entry level,32,None
11,Videographer + Social Media,seeking an independent and detail-oriented soc...,1,Fresh/entry level,"Lighting, Video Editor, Property, Adobe Premie...",1200,3000,4,Monthly,OCC062006,24314,2020v2,1,"Advertising / Media, Design","Contract, Freelance, Internship/Attachment",Fresh/entry level,51,0239


# Save Cleaned Data

In [ ]:
print("\n=== Jobs Validation ===")

print(f"Shape: {df_jobs.shape}")
print(f"Null counts:\n{df_jobs.isnull().sum()}")
print(f"\nKey columns present: {set(['title', 'description', 'skills']).issubset(set(df_jobs.columns))}")

# Data quality check
jobs_no_description = (
    df_jobs['description'].isna() |
    (df_jobs['description'] == '')
).sum()

print(f"\nJobs without description: {jobs_no_description}")

if jobs_no_description > 0:
    print("Removing jobs without descriptions...")
    df_jobs = df_jobs[
        df_jobs['description'].notna() &
        (df_jobs['description'] != '')
    ]
    print(f"Jobs after cleanup: {df_jobs.shape[0]}")

print("\n✓ Jobs validation complete")

\n=== Data Validation ===
\n** Courses Validation **
Shape: (10516, 8)
Null counts:\nuniversity         0
department         0
faculty         2017
code               0
title              0
description        0
course_type    10317
skills             0
dtype: int64
\nData types:\nuniversity     object
department     object
faculty        object
code           object
title          object
description    object
course_type    object
skills         object
dtype: object
\nUnique universities: ['NUS' 'NTU' 'SUTD']
\n** Jobs Validation **
Shape: (7674, 24)
Null counts:\ntitle                        0
description                  0
minimumYearsExperience       0
shiftPattern              7674
skills                       0
otherRequirements         7674
salary_minimum               0
salary_maximum               0
salary_type_id               0
salary_type                  0
occupationId                 0
ssocCode                     0
ssocVersion                  0
numberOfVacancies         

In [ ]:
from pathlib import Path

output_dir = Path("../../data/cleaned")
output_dir.mkdir(exist_ok=True)

df_jobs.to_pickle(output_dir / "jobs_cleaned.pkl")

print(f"Saved jobs data to {output_dir}")
print(f"Jobs: {df_jobs.shape[0]} records")

Saved cleaned data to ../../data/cleaned
Courses: 10516 records
Jobs: 7674 records
